nisapi/clean.py

```diff
@@ -0,0 +1,253 @@
+                        "indicator_type",
+                    ]
+                ).str.to_lowercase()
+            )
+            # cast types
+            .with_columns(
+                pl.col("week_ending").str.strptime(pl.Datetime, "%Y-%m-%dT%H:%M:%S%.f"),
+                pl.col(["estimate", "ci_half_width_95pct"]).cast(pl.Float64),
+            )
+            # change percents to proportions
+            .with_columns(pl.col(["estimate", "ci_half_width_95pct"]) / 100.0)
+        )
+
+        # check that the date doesn't have any trailing seconds
+        assert (clean["week_ending"].dt.truncate("1d") == clean["week_ending"]).all()
+        clean = clean.with_columns(pl.col("week_ending").dt.date())
+
+        clean = (
+            clean
+            # Change from "national" to "nation", so that types are nouns rather
+            # than adjectives. (Otherwise we would need to change "region" to "regional")
+            .with_columns(
+                pl.col("geographic_type").replace({"national": "nation"}),
+                pl.col("geographic_value").replace({"National": "nation"}),
+            )
+            # "sw5n-wg2p" only:
+            # this dataset has an error: for `demographic_type="overall"`, it has
+            # `demographic_value="18+ years"`, but it should be "overall"
+            .with_columns(
+                demographic_value=pl.when(
+                    pl.col("demographic_type") == pl.lit("overall")
+                )
+                .then(pl.lit("overall"))
+                .otherwise(pl.col("demographic_value"))
+            )
+        )
+
+        # there should now be no nulls in any column
+        assert clean.null_count().pipe(sum).item() == 0
+
+        # Remove duplicate rows
+        clean = clean.filter(clean.is_duplicated().not_())
+
+        # Find the rows that are *almost* duplicates: there are some rows that
+        # have nearly duplicate values
+        clean = clean.pipe(remove_near_duplicates)
+
+        # Verify that indicator type "up-to-date" has only one value ("yes")
+        assert clean.filter(pl.col("indicator_type") == pl.lit("up-to-date")).pipe(
+            col_values_in, "indicator_value", ["yes"]
+        )
+
+        # check that "Yes" and "Received a vaccination" are the same thing, so that
+        # we can drop "Up to Date"
+        assert (
+            clean.filter(
+                pl.col("indicator_value").is_in(["yes", "received a vaccination"])
+            )
+            .drop("indicator_type")
+            .pivot(on="indicator_value", values=["estimate", "ci_half_width_95pct"])
+            .select(
+                (
+                    (
+                        pl.col("estimate_yes")
+                        == pl.col("estimate_received a vaccination")
+                    )
+                    & (
+                        pl.col("ci_half_width_95pct_yes")
+                        == pl.col("ci_half_width_95pct_received a vaccination")
```

Comment:
I'm getting a little confused in here: it looks like the phrases "yes" and "received a vaccination" are now part of column names rather than values. That seems wrong, but perhaps I'm just not following the logic.

Reply:
This is what happens when you `.pivot()` with multiple `values` columns. The data have some "indices" (demography, geography, date), one `"on"` column (indicator value, either "yes" or "received a vaccination") and then two value columns (estimate and CI).

So `.pivot()` produces 4 new columns, one for each combination of indicator value and value column name.

This code is kind of a mess and is something that should be generalized and made a cleaning helper, since I suspect this type of question (do these two indicator values mean the same thing?) will arise agian.

Reply:
See #20